smartsim/entity/files.py

```diff
@@ -130,3 +130,122 @@ def _check_path(self, file_path):
         if path.isdir(full_path):
             return full_path
         raise SSConfigError(f"File or Directory {file_path} not found")
+
+
+class TaggedFilesHierarchy:
+    """The TaggedFilesHierarchy class maintains a list of files and a list of
+    child TaggedFilesHierarchy instances. Each instance describes how a
+    directory structure of tagged files.
+
+    TaggedFilesHierarchy represents a directory containing potentially tagged
+    files and subdirectories.
+
+    TaggedFilesHierarchy.base is the directory path from the the root of the
+    file structure
+    E.g.
+    TaggedFilesHierarchy.base = ""            => ./
+    TaggedFilesHierarchy.base = "dir_a"       => ./dir_a
+    TaggedFilesHierarchy.base = "dir_a/dir_b" => ./dir_a/dir_b
+
+    TaggedFilesHierarchy.files is a collection of paths to files that need
+    to be copied to directory that the TaggedFilesHierarchy represents
+
+    TaggedFilesHierarchy.dirs is a collection of child TaggedFilesHierarchy,
+    each with its own files and dirs, representing a directory that needs to
+    be created within the current directory represented by the current
+    TaggedFilesHierarchy
+
+    By performing a depth first search over the entire hierarchy starting at
+    the root (TaggedFilesHierarchy.base = ""), one could reconstruct the
+    tagged file directory structure with relative ease at any location deemed
+    necessary.
+    """
+
+    def __init__(self, base=""):
+        """Initialize a TaggedFilesHierarchy
+
+        :param base: pathlike string specifing the generated directory
```

Comment:
We specify path-like here, but what happens if I pass a ``pathlib.Path`` to this? (Or as a file)

Reply:
Totally see what you saying! I decided I didn't like the passing of strings everywhere to keep track of a directory structure. Going to make a minor tweak to take advantage of the tree and prevent something like this from occuring

Reply:
Changed the `TaggedFilesHierarchy` constructor here: https://github.com/CrayLabs/SmartSim/pull/119/commits/f52e16378e91db0424abcea448653156beb7faa5

Now to create child `TaggedFilesHierarchy`, instead of passing a route to be created that the `Generator` has to trust is correct and a string, now you simply pass in the parent `TaggedFileHierarchy` and the name of a new directory (represented by the new child `TaggedFilesHierarchy`) you want to create in the parent directory (represented by the parent `TaggedFilesHierarchy`).

The name of the new directory cannot be path-like (i.e. must be a string without path separators) and can only be an empty string when the `TaggedFilesHierarchy` has no parent.

The constructor will use this information to create the `TaggedFilesHierarchy.base` attribute. This way when the `Generator` is traversing over the hierarchy, it is guaranteed that all `TaggedFilesHierarchy.base`s are simple path-like strings, and it will never be asked to create a directory that requires intermediate directories to be made.

This should guarantee that a `TaggedFilesHierarchy` cleanly maps to a directory hierarchy and we don't have to worry about users passing in additional types or pointing to existing files or directories.